### PR TITLE
infoschema: implement infoschema table.

### PR DESCRIPTION
**Pull request by coocood**
Creates a information schema specific table that implements table.Table interface.
The rows in this table is generated by `[]*model.DBInfo` on request.

Further reduce memory allocation.

### CODE REVIEW

**Comment by coocood** [comment]
@hanfei1991 @tiancaiamao PTAL

File: infoschema/tables.go [hunk 1]
```diff
@@ -517,7 +519,162 @@ var tableNameToColumns = map[string]([]columnInfo){
...
+
+type infoschemaTable struct {
+	handle *Handle
+	meta   *model.TableInfo
+	cols   []*table.Column
+	rows   [][]types.Datum
+	mu     sync.Mutex
+}
+
+type schemasSorter []*model.DBInfo
```
Comment (hanfei1991) [inline infoschema/tables.go]: add comments for this line.

File: infoschema/tables.go [hunk 2]
```diff
@@ -517,7 +519,162 @@ var tableNameToColumns = map[string]([]columnInfo){
...
+		fullRows = dataForStatistics(dbs)
+	case tableCharacterSets:
+		fullRows = dataForCharacterSets()
+	case tableCollations:
+		fullRows = dataForColltions()
+	case tableFiles:
+	case tableProfiling:
+	case tablePartitions:
+	case tableKeyColumm:
+	case tableReferConst:
```
Comment (hanfei1991) [inline infoschema/tables.go]: do we need a default branch to return error?

File: infoschema/tables.go [hunk 3]
```diff
@@ -517,7 +519,162 @@ var tableNameToColumns = map[string]([]columnInfo){
...
+		cols:   columns,
+	}
+}
+
+type infoschemaTable struct {
+	handle *Handle
+	meta   *model.TableInfo
+	cols   []*table.Column
+	rows   [][]types.Datum
+	mu     sync.Mutex
```
Comment (hanfei1991) [inline infoschema/tables.go]: what is it used for ?

**Comment by tiancaiamao** [comment]
LGTM

File: infoschema/builder.go [hunk 4]
```diff
@@ -212,79 +212,78 @@ func (b *Builder) copySchemaTables(dbName string) {
...
 
-func (b *Builder) initMemorySchemas() error {
-	info := b.is
-	infoSchemaTblNames := &schemaTables{
-		dbInfo: infoSchemaDB,
-		tables: make(map[string]table.Table, len(infoSchemaDB.Tables)),
+func (b *Builder) createSchemaTablesForDB(di *model.DBInfo) error {
+	schTbls := &schemaTables{
+		dbInfo: di,
+		tables: make(map[string]table.Table),
```
Comment (zimulala) [inline infoschema/builder.go]: Add length.

File: infoschema/tables.go (hunk 2 above)
Comment (coocood) [inline infoschema/tables.go]: It's not an error if we don't have any rows for a table.

File: infoschema/builder.go [hunk 5]
```diff
@@ -212,79 +212,78 @@ func (b *Builder) copySchemaTables(dbName string) {
...
 
-// Build sets new InfoSchema to the handle in the Builder.
-func (b *Builder) Build() error {
-	err := b.handle.refillMemoryTables(b.is.AllSchemas())
-	if err != nil {
-		return errors.Trace(err)
+func (b *Builder) createSchemaTablesForInfoSchemaDB() {
+	infoSchemaSchemaTables := &schemaTables{
+		dbInfo: infoSchemaDB,
+		tables: make(map[string]table.Table),
```
Comment (zimulala) [inline infoschema/builder.go]: ditto

File: infoschema/builder.go [hunk 6]
```diff
@@ -212,79 +212,78 @@ func (b *Builder) copySchemaTables(dbName string) {
...
 		tables: make(map[string]table.Table, len(perfSchemaDB.Tables)),
 	}
-	info.schemaMap[perfSchemaDB.Name.L] = perfSchemaTblNames
-
+	b.is.schemaMap[perfSchemaDB.Name.L] = perfSchemaTblNames
 	for _, t := range perfSchemaDB.Tables {
 		tbl, ok := perfHandle.GetTable(t.Name.O)
 		if !ok {
-			return ErrTableNotExists.Gen("table `%s` is missing.", t.Name)
+			continue
```
Comment (coocood) [inline infoschema/builder.go]: We don't care if some performance schema table is missing.

File: executor/builder.go [hunk 7]
```diff
@@ -522,13 +522,14 @@ func (b *executorBuilder) buildTableScan(v *plan.PhysicalTableScan) Executor {
...
-		seekHandle: math.MinInt64,
-		ranges:     v.Ranges,
+		t:            table,
+		asName:       v.TableAsName,
+		ctx:          b.ctx,
+		columns:      v.Columns,
+		schema:       v.GetSchema(),
+		seekHandle:   math.MinInt64,
+		ranges:       v.Ranges,
+		isInfoSchema: v.DBName.L == "information_schema",
```
Comment (zimulala) [inline executor/builder.go]: Add a constant.

**Comment by hanfei1991** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**